ess-r-package improvements

vspinu opened the pull request:
A bunch of improvements aimed to increase devtools command flexibility. Once approved, I would like to add customization variables for each devtools command actions.

vspinu commented:
Hi @lionel-.  Just pinging to make sure that you haven't missed the request. Would you have time to have a look at this in the near future?

lionel- commented:
oops sorry. I was at RStudio conf and am still in San Diego until Tuesday. I'll try to take a look at it next week!

lionel- commented on lisp/ess-r-package.el:
```diff
@@ -321,34 +348,44 @@ the other hand, `.metadata' is always ignored by the R package
-(defun ess-r-devtools-install-package (&optional alt)
-  "Interface to `devtools::install()'."
+  (ess-r-package-eval-linewise
+   "devtools::document(%s)\n" "Documenting %s"
```
This is missing the `arg` argument.

lionel- commented on lisp/ess-r-package.el:
```diff
@@ -310,9 +338,8 @@ filter argument. Otherwise, use the whole base filename."
 ")
 
-(defun ess-r-devtools-revdep-check-package (&optional alt)
+(defun ess-r-devtools-revdep-check-package ()
```
Should you remove `interactive` here?

Maybe we should retire this function altogether, I have never used it and I think it's outdated. I wouldn't be surprised if no one uses it.

vspinu replied:
I haven't used it either and I also don't think it's useful. revdep is a complex procedure which almost always would require a separate script file. So, yes .. let's retire.

vspinu commented on lisp/ess-r-package.el:
```diff
@@ -92,7 +92,6 @@ is searched from that directory instead of `default-directory'."
 
 (defun ess-r-package-name (&optional dir)
   "Return the name of the current package as a string."
-  ;; FIXME Emacs 25.1: Use `when-let'
```
There are some many of such places that it's hardly useful. Plus when-let is deprecated in emacs26 in favor of when-let*.

lionel- replied:
I think it's precisely because there are many such places that it is useful. But point taken for `when-let*`, I wasn't aware of this.

vspinu commented on lisp/ess-r-package.el:
```diff
@@ -321,34 +348,44 @@ the other hand, `.metadata' is always ignored by the R package
+  (ess-r-package-eval-linewise
+   "devtools::install(%s)\n" "Installing %s" arg
+   '("keep_source = TRUE"
+     (read-string "Install args: " "keep_source = TRUE, local = FALSE")
```
When one want to re-build the package from scratch (don't re-use dlls). There is no `cleanup` argument. 

Keep source works in this case if the package is in `ess-r-package-library-paths`.

lionel- replied:
FYI I'm using `load_all(recompile = TRUE)` prior to installing to achieve this.

vspinu replied:
Right, recompile=TRUE is on C-u and local=FALSE is the closest alternative. 

I am staying away from load_all because it didn't work well in the past and it messed with help files a big deal. But maybe I should give it one more chance.

vspinu commented on lisp/ess-r-package.el:
```diff
@@ -321,34 +348,44 @@ the other hand, `.metadata' is always ignored by the R package
+   "devtools::install(%s)\n" "Installing %s" arg
+   '("keep_source = TRUE"
+     (read-string "Install args: " "keep_source = TRUE, local = FALSE")
+     (read-string "Install args: " "keep_source = TRUE, quick = TRUE, upgrade_dependencies = FALSE"))))
```
Ok.

vspinu commented on lisp/ess-r-package.el:
```diff
@@ -242,53 +179,144 @@ Root is determined by locating `ess-r-package-root-file'."
-                              alt "recompile = TRUE"))
+  (ess-r-package-eval-linewise
+   "devtools::load_all(%s)\n" "Loading %s" arg
+   '("" (read-string "Load args:" "recompile = TRUE"))))
```
Good catch.

vspinu commented on lisp/ess-r-package.el:
```diff
@@ -446,6 +483,11 @@ disable the mode line entirely."
+(defvaralias 'ess-r-package-auto-set-evaluation-env 'ess-r-package-auto-enable-namespaced-evaluation)
+(fset 'ess-r-devtools-ask 'ess-r-devtools-execute-command)
+(define-obsolete-function-alias 'ess-r-devtools-ask 'ess-r-devtools-execute-command "18.04")
+(define-obsolete-variable-alias 'ess-r-package-auto-set-evaluation-env ess-r-package-auto-enable-namespaced-evaluation "18.04")
```
Good catch!

vspinu commented on lisp/ess-r-mode.el:
```diff
@@ -115,10 +115,17 @@
+    (define-key ess-r-package-dev-map "e"    'ess-r-devtools-execute-command)
+    (define-key ess-r-package-dev-map "\C-b" 'ess-r-devtools-build)
+    (define-key ess-r-package-dev-map "b"    'ess-r-devtools-build)
+    (define-key ess-r-package-dev-map "\C-B" 'ess-r-devtools-check-package-buildwin)
```
So put it on \C-C then?

lionel- replied:
I let you decide ;)

In my config I'm using `<leader> d c c` for normal checking and `<leader> d c w` for checking on Windows. Using an additional prefix key seems more extensible, i.e. we might want to add common r-hub platforms here.

vspinu replied:
It makes sense but given that basic check is so much more common than all others I am a bit reluctant. If only R-hub worked as expected. I was never able to build on solaris there. 

Two potential alternatives:

  - put all "web" checks on C-c C-w C-w
  - offer alternative check options with completion on C-c C-w C 


BTW, shall we rename the command to `ess-r-package-check-with-winbuilder to avoid mnemonic confusion? The R-hub version could be `ess-r-package-check-with-rhub` with completion for OS types.

lionel- replied:
> put all "web" checks on C-c C-w C-w

This seems reasonable.

>  shall we rename the command to ess-r-package-check-with-winbuilder to avoid mnemonic confusion?

Sounds good.

> with completion for OS types.

That'd be great! The completions can be generated with `rhub::platforms()`.

lionel- replied:
My main use of rhub nowadays is the "ubuntu-rchk" platform btw. It runs Tomas's protection checks.

vspinu commented:
I have incorporated all your suggestions. Added rhub check (on a new C-c C-w C-c ess-r-package-check-map; it's a small price to pay for consistency) and removed revdep check.

lionel- commented:
Thanks!

lionel- commented:
About `ess-r-devtools-execute-command`. I think we should change it to `ess-r-package-execute-command` with the following spec:

- It would change the working directory to the package root temporarily, if not already set.
- It would lookup `ess-r-package-command-packages` which would be set to `(devtools usethis revdepcheck)` by default. The user would be prompted which package the command comes from.
- It would then prompt which command with completion for all exported functions (rather than just functions taking `pkg` arguments like we have now).
- With prefix it'd prompt for additional arguments.

What do you think? This'd be more robust to API changes, more flexible and more useful.

vspinu commented:
Sounds reasonable but I think this command is not that useful (to me anyways). If something is a common pattern it should have a ess-r-package specialized wrapper. Rarely used commands and more complex patterns could be devised at the terminal or in a script. 

> The user would be prompted which package the command comes from.

I think we can skip this part and directly jump to the completions of the form `("build(devtools)", "browse_cran(usethis)" ... etc.

> It would then prompt which command with completion for all exported functions (rather than just functions taking pkg arguments like we have now).

This would likely increase the complexity of the UI. What's the point of having non-pkg functions there? And how will you determine when to plug-in pkg-root and when not?

lionel- commented:
> This would likely increase the complexity of the UI.

No because most package functions use the current directory as default, and it would be temporarily set to the package root. This way we don't have to pass in the pkg root. I think we should do this for most devtools functions.

> What's the point of having non-pkg functions there?

Robustness with different APIs. This will become important with post-diaspora devtools (which introduces backward incompatible changes) and third party packages like revdepcheck.

lionel- commented:
> If something is a common pattern it should have a ess-r-package specialized wrapper. Rarely used commands and more complex patterns could be devised at the terminal or in a script.

I think the completion tool is more useful and I often use it. There are too many commands to give each of them a shortcut.

vspinu commented:
> This way we don't have to pass in the pkg root. I think we should do this for most devtools functions.

This might be be ok for ess-r-package-execute-command but not for main commands. We are printing the command to the REPL, doing so with setwd + tryCatch would be annoying. 

> I think the completion tool is more useful and I often use it.

Which one is that?

lionel- commented:
> We are printing the command to the REPL, doing so with setwd + tryCatch would be annoying.

It should be invisible.

> Which one is that?

The former `ess-r-devtools-ask`.

vspinu commented:
> It should be invisible.

The convention in ESS is to make it clear what commands are run as much as
possible. Users could then check help files customize without being
concerned with the "magic" behind the scene. I wouldn't like to change that
without a good reason. Certainly not for the interactive commands in the
keymap.



On 16 February 2018 at 15:23, Lionel Henry <notifications@github.com> wrote:

> We are printing the command to the REPL, doing so with setwd + tryCatch
> would be annoying.
>
> It should be invisible.
>
> Which one is that?
>
> The former ess-r-devtools-ask.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/emacs-ess/ESS/pull/479#issuecomment-366248418>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABTOC-XOOngfTi01ahjRKQn5g9k5MhzOks5tVY9egaJpZM4R3_2w>
> .
>

lionel- commented:
I agree in general but this is an implementation detail. The working directory will be back to its original state after the command has finished running. This is a background thing.

vspinu commented:
We can in principle display one thing in REPL and run the wrapped version in the background, but then traceback in case of errors might be confusing. 

If you think this is really necessary let's do the setwd magic for commands in ess-r-package-execute-command and leave the interactive commands with the explicit path.  

We should add `.ess_in_pkgwd` witch would do the setting/unseting work.

lionel- commented:
You're right it will be safer and easier to do it from R. We'll need to reimplement package detection on the R side though.

lionel- commented:
@vspinu For installing from github, does it make sense to replace this:

```
     '("" (read-string "Arguments: " "force = TRUE"))
```

by this:

```
     '(""
       (read-string "Arguments: " "quick = TRUE, upgrade_dependencies = FALSE")
       (read-string "Arguments: " "force = TRUE"))
```

This is more consistent with the interface for `install()`. Also we shouldn't use the package mechanism because currently `install_github()` can only be used from within a package which is an artificial restriction.

vspinu commented:
> oes it make sense to replace this:

It does.

> Also we shouldn't use the package mechanism because currently install_github() can only be used from within a package which is an artificial restriction.

That would mean hooking it in ess map. We can add put it in the extra map. C-c C-e g?

lionel- commented:
Agreed.
